File: merge_sort.py
def insertion_sort(ar, len_array):

    for x in range(1, len_array):

        curPos = x

        while curPos > 0 and ar[curPos] < ar[curPos - 1]:
            ar[curPos], ar[curPos - 1] = ar[curPos - 1], ar[curPos]
            curPos -= 1
            print "@"


def merge_sort(ar, left, right):

    if right - left > 1:

        mid = (left + right) / 2

        print '#'

        merge_sort(ar, left, mid)
        merge_sort(ar, mid, right)

        merge(ar, left, mid, right)


def merge(ar, left, mid, right):

    first_half = ar[left:mid]
    second_half = ar[mid:right]

    i = j = 0
    k = left

    while i < mid - left and j < right - mid:

        if first_half[i] <= second_half[j]:
            ar[k] = first_half[i]
            i += 1

        else:
            ar[k] = second_half[j]
            j += 1

        k += 1

    if i != mid - left - 1:
        while i < mid - left:
            ar[k] = first_half[i]
            i += 1
            k += 1

    if j != right - mid - 1:
        while j < right - mid:
            ar[k] = second_half[j]
            j += 1
            k += 1


ar = [6, 3, 8, 10, 5, 6]
print ar
ar1 = ar
insertion_sort(ar1, len(ar))
print ar1
merge_sort(ar, 0, len(ar))

print ar

File: CPP/Algorithms/MergeSort.cpp
#include <iostream>
#include <stdlib.h>
using namespace std;

void merge(int ar[], int left, int right, int mid)
{
    int n1, n2;
    n1 = mid - left + 1;
    n2 = right - mid;

    int temp1[n1], temp2[n2];

    for (int i = 0; i < n1; i++)
        temp1[i] = ar[left + i];
    for (int i = 0; i < n2; i++)
        temp2[i] = ar[mid + i + 1];

    int i = 0, j = 0, k = left;
    while (i < n1 && j < n2)
    {
        if (temp1[i] < temp2[j])
            ar[k++] = temp1[i++];
        else
            ar[k++] = temp2[j++];
    }

    while (i < n1)
        ar[k++] = temp1[i++];
    while (j < n2)
        ar[k++] = temp2[j++];
}

void mergeSort(int ar[], int left, int right)
{
    if (left < right)
    {
        int mid = left + (right - left) / 2;
        mergeSort(ar, left, mid);
        mergeSort(ar, mid + 1, right);
        merge(ar, left, right, mid);
    }
}

// Test Code

int main()
{
    int ar[] = {1, 5, 2, 3, 0, 9};
    int n = sizeof(ar) / sizeof(ar[0]);
    mergeSort(ar, 0, n - 1);
    for (int i = 0; i < n; i++)
        cout << ar[i] << ' ';
    return 0;
}